storage/src/merkle_storage_stats.rs

```diff
@@ -0,0 +1,256 @@
+        self.current += latency;
+    }
+
+    fn end_block(&mut self) {
+        self.latencies.push(self.current);
+        self.current = 0;
+    }
+
+    pub fn get(&self, offset_from_last_applied: usize) -> Option<u64> {
+        self.latencies
+            .len()
+            .checked_sub(offset_from_last_applied + 1)
+            .and_then(|index| self.latencies.get(index))
+            .copied()
+    }
+}
+
+pub enum MerkleStorageAction {
+    Set,
+    Get,
+    GetByPrefix,
+    GetKeyValuesByPrefix,
+    GetContextTreeByPrefix,
+    GetHistory,
+    Mem,
+    DirMem,
+    Copy,
+    Delete,
+    DeleteRecursively,
+    Commit,
+    Checkout,
+    GarbageCollector,
+}
+
+impl fmt::Display for MerkleStorageAction {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            MerkleStorageAction::Set => {
+                write!(f, "Set").unwrap();
+            }
+            MerkleStorageAction::Get => {
+                write!(f, "Get").unwrap();
+            }
+            MerkleStorageAction::GetByPrefix => {
+                write!(f, "GetByPrefix").unwrap();
+            }
+            MerkleStorageAction::GetKeyValuesByPrefix => {
+                write!(f, "GetKeyValuesByPrefix").unwrap();
+            }
+            MerkleStorageAction::GetContextTreeByPrefix => {
+                write!(f, "GetContextTreeByPrefix").unwrap();
+            }
+            MerkleStorageAction::GetHistory => {
+                write!(f, "GetHistory").unwrap();
+            }
+            MerkleStorageAction::Mem => {
+                write!(f, "Mem").unwrap();
+            }
+            MerkleStorageAction::DirMem => {
+                write!(f, "DirMem").unwrap();
+            }
+            MerkleStorageAction::Copy => {
+                write!(f, "Copy").unwrap();
+            }
+            MerkleStorageAction::Delete => {
+                write!(f, "Delete").unwrap();
+            }
+            MerkleStorageAction::DeleteRecursively => {
+                write!(f, "DeleteRecursively").unwrap();
+            }
+            MerkleStorageAction::Commit => {
+                write!(f, "Commit").unwrap();
+            }
+            MerkleStorageAction::Checkout => {
+                write!(f, "Checkout").unwrap();
+            }
+            MerkleStorageAction::GarbageCollector => {
+                write!(f, "GarbageCollector").unwrap();
+            }
+        };
+        Ok(())
+    }
+}
+
+pub struct StatUpdater<'a> {
+    stats: &'a mut MerkleStorageStatistics,
+    timer: Instant,
+    action: MerkleStorageAction,
+    key: Option<String>,
+}
+
+impl<'a> Drop for StatUpdater<'a> {
+    fn drop(&mut self) {
+        self.update_execution_stats()
+    }
+}
+
+impl<'a> StatUpdater<'a> {
+    pub fn new(
+        stats: &'a mut MerkleStorageStatistics,
+        action: MerkleStorageAction,
+        action_key: Option<&Vec<String>>,
+    ) -> Self {
+        let key = match action_key {
```

Comment:
@mat13mn What happens with the rest of the path here?

Shorter but not necessarily better versions:

```rust
    let key = action_key
        .filter(|path| path.len() > 1 && path[0] == "data")
        .map(|path| path[1].to_string());
```

or

```rust
    let key = match action_key {
        Some(path) if path.len() > 1 && path[0] == "data" => Some(path[1].to_string()),
        Some(_) | None => None,
    };
```

Reply:
>@mat13mn What happens with the rest of the path here?

the idea was to store information about the first key only - something like `max-depth=1` parameter for `find`